graylog2-server/src/main/java/org/graylog/failure/FailureBatch.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.failure;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * A structure bearing a batch of failures. It guarantees all failures
+ * to be instances of the same class
+ */
+public class FailureBatch {
+
+    private final List<? extends Failure> failures;
+    private final Class<? extends Failure> failureClass;
+
+    private FailureBatch(List<? extends Failure> failures,
+                        Class<? extends Failure> failureClass) {
+        this.failures = ImmutableList.copyOf(failures);
+        this.failureClass = failureClass;
+        this.failures.forEach(f -> {
+            if (!f.getClass().equals(failureClass)) {
+                throw new IllegalArgumentException("Not all failures from the list are instances of " + failureClass.getName());
+            }
+        });
```

Comment:
Wouldn't that be a good candidate to use generics and let the compiler make sure that all failures in the batch are of the same type? E.g. make it

```
public class FailureBatch<T extends Failure>
```

Reply:
So at first I liked the idea and I implemented it.
But looking at the diff, I'm not so sure anymore..
Lots of clutter for not much gain. We only provide two static creator methods, and they only accept
the specific failure types.
So :shrug: I could live without generics, in this case..

Here's part of the diff (no enterprise)
```
diff --git graylog2-server/src/main/java/org/graylog/failure/DefaultFailureHandler.java graylog2-server/src/main/java/org/graylog/failure/DefaultFailureHandler.java
index 20db90b556..6248b65726 100644
--- graylog2-server/src/main/java/org/graylog/failure/DefaultFailureHandler.java
+++ graylog2-server/src/main/java/org/graylog/failure/DefaultFailureHandler.java
@@ -36,7 +36,7 @@ public class DefaultFailureHandler implements FailureHandler {
     }
 
     @Override
-    public void handle(FailureBatch failureBatch) {
+    public void handle(FailureBatch<? extends Failure> failureBatch) {
         failureBatch.getFailures().forEach(failure ->
                 indexFailureService.saveWithoutValidation(new IndexFailureImpl(ImmutableMap.<String, Object>builder()
                         .put("letter_id", failure.failedMessage().getId())
@@ -48,7 +48,7 @@ public class DefaultFailureHandler implements FailureHandler {
     }
 
     @Override
-    public boolean supports(FailureBatch failureBatch) {
+    public boolean supports(FailureBatch<? extends Failure> failureBatch) {
         return failureBatch.containsIndexingFailures();
     }
 
diff --git graylog2-server/src/main/java/org/graylog/failure/FailureBatch.java graylog2-server/src/main/java/org/graylog/failure/FailureBatch.java
index d4668828df..c58a670cc7 100644
--- graylog2-server/src/main/java/org/graylog/failure/FailureBatch.java
+++ graylog2-server/src/main/java/org/graylog/failure/FailureBatch.java
@@ -25,13 +25,13 @@ import java.util.List;
  * A structure bearing a batch of failures. It guarantees all failures
  * to be instances of the same class
  */
-public class FailureBatch {
+public class FailureBatch<T extends Failure> {
 
-    private final List<? extends Failure> failures;
-    private final Class<? extends Failure> failureClass;
+    private final List<T> failures;
+    private final Class<T> failureClass;
 
-    private FailureBatch(List<? extends Failure> failures,
-                        Class<? extends Failure> failureClass) {
+    private FailureBatch(List<T> failures,
+                         Class<T> failureClass) {
         this.failures = ImmutableList.copyOf(failures);
         this.failureClass = failureClass;
         this.failures.forEach(f -> {
@@ -46,8 +46,8 @@ public class FailureBatch {
      * @return a batch of indexing failures
      * @throws IllegalArgumentException if not all failures are instances of {@link IndexingFailure}
      */
-    public static FailureBatch indexingFailureBatch(List<IndexingFailure> indexingFailures) {
-        return new FailureBatch(indexingFailures, IndexingFailure.class);
+    public static FailureBatch<IndexingFailure> indexingFailureBatch(List<IndexingFailure> indexingFailures) {
+        return new FailureBatch<>(indexingFailures, IndexingFailure.class);
     }
 
     /**
@@ -55,24 +55,25 @@ public class FailureBatch {
      * @return a batch of processing failures
      * @throws IllegalArgumentException if not all failures are instances of {@link ProcessingFailure}
      */
-    public static FailureBatch processingFailureBatch(List<ProcessingFailure> processingFailures) {
-        return new FailureBatch(processingFailures, ProcessingFailure.class);
+    public static FailureBatch<ProcessingFailure> processingFailureBatch(List<ProcessingFailure> processingFailures) {
+        return new FailureBatch<>(processingFailures, ProcessingFailure.class);
     }
 
     /**
      * Creates a batch containing only one processing failure
+     *
      * @param processingFailure a processing failure to include in this batch
      * @return a batch with the processing failure
      * @throws IllegalArgumentException if the failure is not instances of {@link ProcessingFailure}
      */
-    public static FailureBatch processingFailureBatch(ProcessingFailure processingFailure) {
+    public static FailureBatch<ProcessingFailure> processingFailureBatch(ProcessingFailure processingFailure) {
         return processingFailureBatch(ImmutableList.of(processingFailure));
     }
 
     /**
      * @return a list of failures within the batch. The returned collection is immutable.
      */
-    public List<? extends Failure> getFailures() {
+    public List<T> getFailures() {
         return failures;
     }
 
@@ -106,8 +107,12 @@ public class FailureBatch {
 
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
-        if (o == null || getClass() != o.getClass()) return false;
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
         final FailureBatch that = (FailureBatch) o;
         return Objects.equal(failures, that.failures) && Objects.equal(failureClass, that.failureClass);
     }
diff --git graylog2-server/src/main/java/org/graylog/failure/FailureHandler.java graylog2-server/src/main/java/org/graylog/failure/FailureHandler.java
index f68b26893d..41cb645036 100644
--- graylog2-server/src/main/java/org/graylog/failure/FailureHandler.java
+++ graylog2-server/src/main/java/org/graylog/failure/FailureHandler.java
@@ -34,7 +34,7 @@ public interface FailureHandler {
      *
      * @param failureBatch a batch of failures, supported by this handler
      */
-    void handle(FailureBatch failureBatch);
+    void handle(FailureBatch<? extends Failure> failureBatch);
 
     /**
      * Guides the master failure handling service, when deciding
@@ -43,7 +43,7 @@ public interface FailureHandler {
      * @param failureBatch a batch of failures to test
      * @return true if the batch can be processed by this handler
      */
-    boolean supports(FailureBatch failureBatch);
+    boolean supports(FailureBatch<? extends Failure> failureBatch);
 
     /**
      * Guides the master failure handling service, when deciding
diff --git graylog2-server/src/main/java/org/graylog/failure/FailureHandlingService.java graylog2-server/src/main/java/org/graylog/failure/FailureHandlingService.java
index 67398eed2a..0f9f720b31 100644
--- graylog2-server/src/main/java/org/graylog/failure/FailureHandlingService.java
+++ graylog2-server/src/main/java/org/graylog/failure/FailureHandlingService.java
@@ -127,7 +127,7 @@ public class FailureHandlingService extends AbstractExecutionThreadService {
         logger.debug("The service has been interrupted.");
     }
 
-    private void handle(FailureBatch failureBatch) {
+    private void handle(FailureBatch<? extends Failure> failureBatch) {
         suitableHandlers(failureBatch)
                 .forEach(handler -> {
                     try {
@@ -149,7 +149,7 @@ public class FailureHandlingService extends AbstractExecutionThreadService {
         }
     }
 
-    private List<FailureHandler> suitableHandlers(FailureBatch failureBatch) {
+    private List<FailureHandler> suitableHandlers(FailureBatch<? extends Failure> failureBatch) {
         final List<FailureHandler> suitableHandlers = suitableHandlers(failureHandlers, failureBatch)
                 .filter(FailureHandler::isEnabled)
                 .collect(Collectors.toList());
@@ -162,7 +162,7 @@ public class FailureHandlingService extends AbstractExecutionThreadService {
         }
     }
 
-    private Stream<FailureHandler> suitableHandlers(Collection<FailureHandler> handlers, FailureBatch failureBatch) {
+    private Stream<FailureHandler> suitableHandlers(Collection<FailureHandler> handlers, FailureBatch<? extends Failure> failureBatch) {
         return handlers.stream()
                 .filter(h -> h.supports(failureBatch));
     }
diff --git graylog2-server/src/main/java/org/graylog/failure/FailureSubmissionQueue.java graylog2-server/src/main/java/org/graylog/failure/FailureSubmissionQueue.java
index afe07e9c0a..45eed629b8 100644
--- graylog2-server/src/main/java/org/graylog/failure/FailureSubmissionQueue.java
+++ graylog2-server/src/main/java/org/graylog/failure/FailureSubmissionQueue.java
@@ -47,7 +47,7 @@ class FailureSubmissionQueue {
 
     private final Logger logger = LoggerFactory.getLogger(getClass());
 
-    private final BlockingQueue<FailureBatch> queue;
+    private final BlockingQueue<FailureBatch<? extends Failure>> queue;
     private final Configuration configuration;
     private final Meter submittedFailureBatches;
     private final Meter submittedFailures;
@@ -98,8 +98,8 @@ class FailureSubmissionQueue {
      * @return one batch from the queue. If the queue is empty,
      * waits for a batch to become available.
      */
-    FailureBatch consumeBlocking() throws InterruptedException {
-        final FailureBatch fb = queue.take();
+    FailureBatch<? extends Failure> consumeBlocking() throws InterruptedException {
+        final FailureBatch<? extends Failure> fb = queue.take();
         consumedFailureBatches.mark();
         consumedFailures.mark(fb.size());
         return fb;
diff --git graylog2-server/src/test/java/org/graylog/failure/FailureSubmissionServiceTest.java graylog2-server/src/test/java/org/graylog/failure/FailureSubmissionServiceTest.java
index c3c0039e3c..1d2b54a130 100644
--- graylog2-server/src/test/java/org/graylog/failure/FailureSubmissionServiceTest.java
+++ graylog2-server/src/test/java/org/graylog/failure/FailureSubmissionServiceTest.java
@@ -38,7 +38,7 @@ public class FailureSubmissionServiceTest {
     private final FailureHandlingConfiguration failureHandlingConfiguration = Mockito.mock(FailureHandlingConfiguration.class);
     private final FailureSubmissionService underTest = new FailureSubmissionService(failureSubmissionQueue, failureHandlingConfiguration);
 
-    private final ArgumentCaptor<FailureBatch> failureBatchCaptor = ArgumentCaptor.forClass(FailureBatch.class);
+    private final ArgumentCaptor<FailureBatch<? extends Failure>> failureBatchCaptor = ArgumentCaptor.forClass(FailureBatch.class);
 
     @Test
     public void submitIndexingErrors_allIndexingErrorsTransformedAndSubmittedToFailureQueue() throws Exception {

```